### DISCUSSION TITLE
how to get the mean value of a folder of images

### DISCUSSION THREAD

**Discussion by TLmv** [ImageMagick/ImageMagick#6413] [Digital Image Processing] [answered]
windows10.IrfanView 7.1.1-11-q16-hdri-x64

i have a folder with images, eg IMG_0001.JPG and up.
what i want to achieve is increase or reduce exposure on some of the images based on the median value of another image....if that is the wrong path, tell me.

because i don't know yet how to do that i first wanted to create a text file with the found median value, so i can process that later

however i'm unexperienced with ImageMagick and unsure how to make this happen in commandline, 

in cmd i tried:
magick identify IMG_0001.JPG
IMG_0001.JPG JPEG 100x100 100x100+0+0 8-bit sRGB 15001B 0.000u 0:00.000
so far so good...then:
magick MagicBooleanType GetImageMean(const Image *IMG_0001.JPG,double *mean,double) 
but that just gave me 'no such file or directory'

after some searching i also tried this:
magick convert IMG_0001.JPG -colorspace gray -depth 8 -define histogram:unique-colors=false histogram:info:- | sed -n 's/^ *[0-9]*: (\([^)]*\)).*$/\1/p' | sort -n | statistic median > IMG_0001.txt
'sed' is not recognized as an internal or external command,
operable program or batch file.

but i'm guessing sed is not part of windows.
I'm sure i got it wrong but could not find an example, could somebody help me out with a piece of code how to make this work and preferrably save the result f.i. same as filename e.g. IMG_****.TXT

**Comment by snibgo** [comment]
What is your version of IM? What is your platform (Windows 11, or whatever)?

`sed` is not supplied with Windows. However, it is available for Windows, for example from Cygwin. See [IM with Cygwin](http://im.snibgo.com/cygwin.htm)

I'm not sure what you mean by "the median value of another image".

I suppose that would be three values, one each for the RGB channels. The value for the red channel would be the value for which 50% of the pixels have a lower value, and 50% of the pixels have a higher value. Similarly for the green and blue channels. Is that what you want?

Then you could find the median of any other image, and multiply its pixel values by `this_median/reference_median` to normalise the images.  Instead of multiplying (which can cause clipping) you might prefer a power transformation.

**Comment by fmw42** [comment]
Please always supply your IM version and platform (as asked by the form) when asking questions. It helps us since we have two main versions of IM and syntax may be different for different platforms.

**Comment by TLmv** [comment]
I'm using the latest version 7.1.1 and windows10 for this task
mean value as the combined rgb values average...i thought that would be similar to brightness or exposure.
but if it has to be 3 separate values that is fine with me.

however, how to get those values?

and yes, using this_mean/reference_mean sounds about what i'm trying to do.

**Comment by snibgo** [comment, marked as answer]
A mean is not the same as a median. See: https://en.wikipedia.org/wiki/Median and https://en.wikipedia.org/wiki/Mean

Finding the mean of an image is simple: "`-scale 1x1!`" reduces the image to one pixel that has the means of the RGB channels. Windows syntax:
```
magick in.png -scale "1x1^!" txt:
```
That gives text values of the means of each channel.

**Reply by TLmv** [reply]
> A mean is not the same as a median. See: https://en.wikipedia.org/wiki/Median and https://en.wikipedia.org/wiki/Mean
> 
sorry, english isn't my first language, for me mean and median looked similar, but now i understand where i went wrong.


> Finding the mean of an image is simple: "`-scale 1x1!`" reduces the image to one pixel that has the means of the RGB channels. Windows syntax:
> 
> ```
> magick in.png -scale "1x1^!" txt:
> ```
> 

that gave me:
ImageMagick pixel enumeration: 1,1,0,255,srgb
0,0: (43,38,33)  #2B2621  srgb(16.9167%,14.8735%,12.9816%)
reference image:
ImageMagick pixel enumeration: 1,1,0,255,srgb
0,0: (49,47,41)  #312F29  srgb(19.326%,18.2451%,16.0659%)

> That gives text values of the means of each channel.

thank you, I'm assuming that is 43,38,33 and 49,47,41 in this case.
i have to sort out how to strip those numbers from the outputfile first before i can take a go at resampling an image with these values.  with use of this_mean/reference_mean

**Comment by fmw42** [comment]
You can get the mean of the intensity of the image (simple average of each channel) in range 0 to 255 as

`magick in.png -intensity average -format "%[fx:255*intensity]\n" info:`

You can get other weighted averages using other -intensity options.  See https://imagemagick.org/script/command-line-options.php#intensity

Alternate to snibgo's answer would be

`magick in.png -scale "1x1^!" -format "%[pixel: u.p{0,0}]\n" info:`

or

`magick in.png -scale "1x1^!" -format "%[fx:u*255]\n" info:`

**Reply by TLmv** [reply]
>magick in.png -intensity average -format "%[fx:255*intensity]\n" info:

gives 33.3333
while
>magick in.png -scale "1x1^!" -format "%[pixel: u.p{0,0}]\n" info:
>or
>magick in.png -scale "1x1^!" -format "%[fx:u*255]\n" info:

gives 49.2812

so i'm guessing i'd have to stick with the second value, however would resampling with one value instead of 3 separate values create the same result? 
i think snibgo was suggesting resampling per channel would be better.

**Reply by fmw42** [reply]
If your image has transparency, then -scale 1x1! will ignore the transparent pixels when doing the average.  There are many ways to get a mean value and the weights per channel will affect the result. Also which channels are involved or if the image is converted to gray, which also has many ways of doing that.

Can you specify a formula you want per channel in combining the channels to form gray. Then just get the mean of the grayscale image.

You can also combine the channels either before or after getting the mean (average).

Please identify what you had in mind

**Comment by snibgo** [comment]
> i think snibgo was suggesting resampling per channel would be better.

I don't really know what you want to do. You might convert to gray, and find the mean of those values. Or find the mean of each channel, and convert that to grayscale. Or do something else.

**Reply by TLmv** [reply]
i have image sets, but they do not always match in exposure. 

what i'm trying to do is batch alter the brightness of the lightest picture to match the darkest....or maybe the other way 'round if it becomes too dark ;)
so my idea is: 
get mean values for both pictures, grab the lightest picture of the 2 (highest values) and use this_mean/reference mean to darken the brightest picture to match the darkest.

**Reply by fmw42** [reply]
Snibgo already has a script to do that on Windows and I have one for Unix systems.

**Reply by TLmv** [reply]
is that so, Snibgo would you care to share a link?.

**Comment by snibgo** [comment]
See my page [Setting the mean](http://im.snibgo.com/setmean.htm) which shows how we can tweak an image to match the brightness (and colour) of another.

**Reply by TLmv** [reply]
thanks, i will. i think negative gain will be my first try, it sounds like the one that fits my goal.
for now i guess, case closed :)
